src/mod/endpoints/mod_sofia/sofia.c

```diff
@@ -3816,6 +3816,8 @@ static void parse_gateways(sofia_profile_t *profile, switch_xml_t gateways_tag,
 	sofia_gateway_t *gp;
 	switch_memory_pool_t *pool;
 	switch_status_t status = SWITCH_STATUS_SUCCESS;
+	
+	sofia_reg_truly_del_gateway(profile);
```

Comment:
This will not lock gw_mutex. Problem is what if in one of the thread is modifying the gateway and you clean it up. This will cause undefined behavior.

Reply:
totally agree. Added gw_mutex lock